src/opam.nix

```diff
@@ -382,6 +382,77 @@ in rec {
+  # deduplicates sources, and provides a list of sources to fetch
+  defsToSrcs = defs:
+    let
+      urlToName = url:
+        let
+          components = lib.strings.splitString "/" url;
+          filename = lib.last components;
+          name = builtins.head (lib.strings.splitString "." filename);
+        in name;
+      defToSrc = { version, ... }@pkgdef:
+        let
+          inherit (getUrl bootstrapPackages pkgdef) src;
+          name = if pkgdef ? dev-repo then urlToName pkgdef.dev-repo else "";
+        in
+        { inherit name version src; };
+    in (lib.attrsets.mapAttrsToList (pkgName: def: defToSrc def) defs);
+  
+  deduplicateSrcs = srcs:
+    # This is O(n^2). We could try and improve this by sorting the list on name. But n is small.
+    let op = srcs: newSrc:
+      # Find if two packages come from the same dev-repo.
+      # Note we are assuming no dev-repos will have different names here, but we also assume
+      # this later when we will symlink in the duniverse directory based on this name.
+      let duplicateSrc = lib.lists.findSingle (src: src.name == newSrc.name) null "multiple" srcs; in
+      # Multiple duplicates should never be found as we deduplicate on every new element.
+      if duplicateSrc == "multiple" then throw "Something has gone terribly wrong"
+      else if duplicateSrc == null then srcs ++ [ newSrc ]
+      # > If packages from the same repo were resolved to different URLs, we need to pick
+      # > a single one. Here we decided to go with the one associated with the package
+      # > that has the higher version. We need a better long term solution as this won't
+      # > play nicely with pins for instance.
+      # > The best solution here would be to use source trimming, so we can pull each individual
+      # > package to its own directory and strip out all the unrelated source code but we would
+      # > need dune to provide that feature.
+      # See [opam-monorepo](https://github.com/tarides/opam-monorepo/blob/9262e7f71d749520b7e046fbd90a4732a43866e9/lib/duniverse.ml#L143-L157)
+      else if duplicateSrc.version >= newSrc.version then srcs
+      else (lib.lists.remove duplicateSrc srcs) ++ [ newSrc ];
+    in builtins.foldl' op [ ] srcs;
+
+  mkSrcsScope = srcs:
+    let
+      # derivation that fetches the source
+      mkSrc = { name, version, src }:
+        bootstrapPackages.pkgsBuildBuild.stdenv.mkDerivation ({
+          inherit name version src;
+          phases = [ "unpackPhase" "installPhase" ];
+          installPhase = ''
+            mkdir $out
+            cp -R . $out
+          '';
+        });
+    in
+    makeScope callPackageWith (self: builtins.listToAttrs (builtins.map
+      (src: lib.attrsets.nameValuePair src.name (mkSrc src)) srcs));
+
+  queryToScopeMonorepo = { repos ? [ opamRepository ], overlays ? [ ], resolveArgs ? {
+        # pass monorepo = 1 to `opam admin list` to pick up dependencies marked with {?monorepo}
+        env.monorepo=1;
```

Comment:
Why not move this down to the `opamList` call itself, so that providing a `resolveArgs` argument doesn't override this setting (which is, AFAIU, always useful for `queryToScopeMonorepo`)?

Reply:
Very good point. I'll implement this.